### ISSUE TITLE
Logging structured data without it appearing in the text message

### ISSUE THREAD

**Issue by roji** [aspnet/Logging#533]
Am looking for some guidance here, or at least confirmation that I understand things right...

Currently, if I want to use the extension methods (e.g. `LogDebug`), it seems that in order to log some piece of structured data it must be referenced from the message - this seems problematic in several scenarios.

First, when using a logging framework such as NLog, it is standard practice to leave the layout of the text message to NLog (user-configurable). Requiring me to embed all structured data in the message text hands off a fully-formed message to NLog, stripping it of one of it's main responsibilities (layouting).

Also, in many scenarios there's a lot of structured context that may be interesting to users in some cases, but that doesn't mean it's *always* interesting and should get included in the text of each and every message. This causes very large log messages with clutter and info nobody really wants (although people may still want the option of looking up these details in the structured data attached to the message).

I understand that I don't have to use the extension methods and can roll my own, but I'd like confirmation to the above, and also your opinion on this.

**Comment by herecydev** [comment]
> First, when using a logging framework such as NLog, it is standard practice to leave the layout of the text message to NLog (user-configurable). Requiring me to embed all structured data in the message text hands off a fully-formed message to NLog, stripping it of one of it's main responsibilities (layouting).

I would probably make a custom `ILogger`, as you're allowed to [give it any TState](https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/ILogger.cs#L22) and let that decide on the format (can be user-configurable), it can even enhance the messages with additional information (current time, machine name, etc) so that the caller doesn't need to supply it.

> Also, in many scenarios there's a lot of structured context that may be interesting to users in some cases, but that doesn't mean it's always interesting and should get included in the text of each and every message. This causes very large log messages with clutter and info nobody really wants (although people may still want the option of looking up these details in the structured data attached to the message).

This one's a little weird and I'm not sure I totally get it. But I think once again, you can switch on the `LogLevel` within your custom logger and decide what's appropriate. With that said, surely the calling code has a much better understanding of verbosity than a downstream consumer. You should really use the `LogLevel` to filter out what's too noisy; a syslog driver will probably log everything whilst a console just information and above.

I see it as the calling code provides as much meaningful data as it can, the logger's can optionally enhance that with additional data but ultimately filter that down based on the appropriate destination. Bearing in mind, we have technologies like Elastic Search to perform more advanced features.

**Comment by roji** [comment]
>>First, when using a logging framework such as NLog, it is standard practice to leave the layout of the text message to NLog (user-configurable). Requiring me to embed all structured data in the message text hands off a fully-formed message to NLog, stripping it of one of it's main responsibilities (layouting).
> I would probably make a custom ILogger, as you're allowed to give it any TState and let that decide on the format (can be user-configurable), it can even enhance the messages with additional information (current time, machine name, etc) so that the caller doesn't need to supply it.

I'm not sure that a custom logger is necessary - the provided Logger's Log method already accepts any TState and formatter... So this is definitely resolvable, it's just not compatible with the "standard" extension methods and requires implementing a layer for something that seems pretty basic/standard.

> This one's a little weird and I'm not sure I totally get it. But I think once again, you can switch on the LogLevel within your custom logger and decide what's appropriate. 

The issue isn't what log *events* should be logged - which is what LogLevel is about - but rather what extra information is included in the textual message that's emitted in the event.

> With that said, surely the calling code has a much better understanding of verbosity than a downstream consumer. You should really use the LogLevel to filter out what's too noisy; a syslog driver will probably log everything whilst a console just information and above.

I don't think that's necessarily true... Coming from NLog, it's quite frequent for applications to log quite a thin textual message, but to include lots of contextual information with it. The calling code should *make available* the information, but whether the textual message needs to include it seems more like a user configuration choice. This allows you to search for log records (in a tool like Kibana) without making the message itself heavier. 

The point is that there are two "usage patterns" for consuming logs. On the one hand, you want to be able to search/filter for logs - this is where the structured information is important. On the other hand, you want to be able to simply browse - read records line-by-line, say around a timestamp or some event - and at that point the message text (and its terseness!) is important. It seems the current approach leads to very heavy messages as every piece of information you *may* want to search on also needs to get included in the text.

> I see it as the calling code provides as much meaningful data as it can, the logger's can optionally enhance that with additional data but ultimately filter that down based on the appropriate destination. Bearing in mind, we have technologies like Elastic Search to perform more advanced features.

I think that's different - enriching an application-provided event with generic information (thread/process id, machine name...) is very different from fields provided by the application itself. In this issue I'm talking about the latter only.

**Comment by herecydev** [comment]
> The calling code should make available the information, but whether the textual message needs to include it seems more like a user configuration choice.

I'll preface this with I'm not an expert, nor have I used NLog. Given the following scenario, where some command failed and you want to log the failure, the command that was used and the exception that was raised:

```c#
catch(Exception ex) {
  _logger.LogInformation("DB Command failed {0} {1}", command, ex);
}
```

You could have a file logger that adds on the current date, formats the message with the parameters to produce something like: `"14/12/2016 | DB Command failed | Select * from foo | foo does not exist"`.
You could then equally have a logger that adds on the current time but drops the parameters for terseness: `"9:20:16 | DB Command failed"`

Does that satisfy the 2 "ends of the spectrum"? Or am I missing the use case? Also I don't think the DI/Logging libraries are meant to be prescriptive, could you use the [NLog extensions](https://github.com/NLog/NLog.Extensions.Logging) to have the behavior you're after working with the `ILogger` abstractions?

**Comment by roji** [comment]
Your example of prefacing the date is what I referred to earlier as "generic information" that can indeed be added by any threading package - this kind of info includes thread ID, hostname or anything else that doesn't need to come from the application. You're right that this case isn't an issue.

The kind of info I'm talking about is stuff that only the application knows about. For example, an event may have info such as server IP, port, transaction ID, etc. These can't be added by NLog (or any other package), because only your application knows about them. On the other hand, it doesn't necessarily make sense to include them in the *text* of each and every log message we log. So we want the information to be available in case want to search for, say, all events we've had with a certain server IP; on the other hand, shoving the server IP into each and every log text is unnecessary and makes logs harder to understand.

In NLog (and other logging packages), the application would log a text message, in addition to the bundle of structured data. The user's NLog configuration would *render* all this information: a user could decide to prefix every message with the current timestamp (generic info), but also with the server IP, the port, or any other info in the structured bundle.

I hope I'm making sense and that the use case is more or less clear...

**Comment by herecydev** [comment]
I think so, you want the application to supply as much information to the logger as possible and for the logger to be configured to either be verbose/terse depending on the situation.

> You could have a file logger that adds on the current date, formats the message with the parameters to produce something like: "14/12/2016 | DB Command failed | Select * from foo | foo does not exist".
You could then equally have a logger that adds on the current time but drops the parameters for terseness: "9:20:16 | DB Command failed"

Why wouldn't this satisfy your requirements? You could have a syslog logger for the first example and a console logger for the latter? The formatting/discarding of information that can happen in the logger could easily be user-configurable (DI in some templates), this can even be done at a per-class level because we have `ILogger<T>`.

**Comment by roji** [comment]
> I think so, you want the application to supply as much information to the logger as possible and for the logger to be configured to either be verbose/terse depending on the situation.

Right, configured by the user.

> Why wouldn't this satisfy your requirements? You could have a syslog logger for the first example and a console logger for the latter? The formatting/discarding of information that can happen in the logger could easily be user-configurable (DI in some templates), this can even be done at a per-class level because we have ILogger<T>.

Because currently it seems that it's impossible to use the extension methods to pass structured information without *also* including it in the message text... The extension methods parse the message text, looking for named variables, and matching those to the params array. So when using the "standard" extension methods, all structured data must also be included in the message text.

**Comment by herecydev** [comment]
I'm with you, I think it boils down to "this works for most people". The default extensions use the `FormattedLogValues` but I don't think that will work as the original message is private. So you wouldn't be able to selectively remove things. Would probably have to create another extension like `LogStructureInformation` that passes through the message format and the args. Then in combination with customer loggers or Nlog you should be able to go from there.

I think if [this was public](https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/Internal/FormattedLogValues.cs#L21) you would be fine.

**Comment by roji** [comment]
Sorry for disappearing :)

Making `_originalMessage` accessible is a possible approach. Note that it's also possible to simply avoid the extension methods and call `Logger.Log()` directly, controlling exactly what state object is passed and how it's formatted.

So users aren't exactly *limited*, it's just that the default way of doing things is basically very inspired from Serilog, and doesn't really take into account other approaches such as NLog's.

**Comment by herecydev** [comment]
> Note that it's also possible to simply avoid the extension methods and call Logger.Log() directly, controlling exactly what state object is passed and how it's formatted.

It is, but telling people not to use these nice, convenient extension `LogInformation`, `LogError` etc and instead use `Log(LogLevel.Information, 0, new FormattedLogValues(message, args), null, (m, e) => m.ToString())` is far from intuitive. It would be preferable to make `FormattedLogValues` have that public property and then people can start extending on it.

**Comment by roji** [comment]
Looking at it again, I'm not sure I understand how making this property public addresses this issue. If I understand correctly, all that would do would be to provide access to the original format string, the one with names in it. Also, it's not clear at what point users would be able to interact with this property: `FormattedLogValues` is created internally when `LogXXX` and immediately passed down to `Logger.Log`... Can you provide more detail on how this would work?

**Comment by herecydev** [comment]
Because in your custom `ILogger` you would be have access to: `"DB Command failed {0} {1}` *and* `Select * from foo` and `foo does not exist`. You've got everything you need to either log everything or strip some of the more verbose things out. I don't envisage users interacting with neither the `FormattedLogValues` or `Log(a, b, c, d, e)` directly. Just using the extensions and letting the different loggers do all the hard work.

**Comment by roji** [comment]
I don't think that solves it... If you continue using the standard extension methods, you continue writing code like this:

```c#
logger.LogDebug(eventId, exception, "Executing: {Sql}", sql, transactionId, someOtherInterestingBit);
```

Your custom `ILogger` indeeds get both the original string (`Executing: {Sql}`) and the processed string (`Executing: {0}`). `transactionId` and `someOtherInterestingBit` are there in the params object array, but there's nothing saying how to name them. The point is that  the extension methods depend on the formatter string to reference (i.e. name) all structural data.

In other words, the problem seems to be with the extension methods' API (which follows Serilog) and does not allow fo structured data that's not referenced in the format string. An alternate API would simply allow passing in a Dictionary, which would represent the structured data. The formatter string would still contain named references to the Dictionary's keys (i.e. names), and integrate their corresponding values in the textual message. But nothing would prevent us from having *additional* key/value pairs in the Dictionary which *aren't* referenced in the format string.

**Comment by herecydev** [comment]
That's correct, this was on the assumption that they were passed in the correct order and with the correct amount. Which probably holds true for most cases. An overloaded extension with a dictionary would be ideal though.

**Comment by roji** [comment]
@herecydev I don't think so... the point is that with the extension methods there *is no way* to name a structured parameter *without* referencing it from the format string... It's not a question of correct order - no matter how you implement your ILogger, with the extension methods you get a string with name references, and a nameless array of objects. The first objects in the array are named via the string - and also appear in the formatted text message - while the remaining ones are completely anonymous and therefore can't really be part of the structured data logged alongside the message.

**Comment by herecydev** [comment]
This was assuming you were using this:
```c#
catch(Exception ex) {
  _logger.LogInformation("DB Command failed {0} {1}", command, ex);
}
```
where you aren't using named references. Otherwise if you need the named references, I only see the dictionary being the working solution.

**Comment by herecydev** [comment]
I actually think this is possible: See [this line](https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/Internal/FormattedLogValues.cs#L45) shows that you can get the `OriginalFormat` out of the values collection. Then using  [these lines](https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/Internal/FormattedLogValues.cs#L65-L71) you can enumerate through the named values thanks to [this class](https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs).

It does therefore look like with the existing extension methods, and your own custom loggers, you have the potential, through DI'ing the templates, to do what you're asking.

**Comment by roji** [comment]
I might not be explaining myself clearly...

My problem is the very fact that your string must contain named reference to the positional parameter list (which, by their nature, are otherwise unnamed). What you're showing allows me to access values that have been named by the string. What I'd like to do is include structured values - with names I choose - *without* referencing them in the format string. That seems impossible to do without adding an overload that accepts a Dictionary.

**Comment by herecydev** [comment]
If I understand what you're after then no it won't be possible at all. You can't "know" what the parameters are at the logger without either naming them in the format string (and thereby deducing through order), or passing KVPs through.

**Comment by roji** [comment]
Yeah, that's my understanding as well. I opened this issue to propose that Microsoft.Extensions.Logging  provide a solution for that, e.g. via overloads that accept that accept a Dictionary.

**Comment by roji** [comment]
In other words, if the team feels like it's something you need to have, I could submit a PR.

**Comment by herecydev** [comment]
Cool... we got there :) It sounds like an interesting addition.

**Comment by Ralstlin** [comment]
@roji I come here looking for exactly the same thing, I was trying to create a loggerProvider that use MongoDB driver to log structured data into mongo, sadly I found the same as you, logging params without naming them in the message like:  `_logger.LogTrace("Creating new user", UserData, SystemData, WhatEver);` will not display these params in the Ilogger.Log because LogValuesFormatter.GetValue() https://github.com/aspnet/Logging/blob/c3f04e4b6f5413d93b73e5519d2674dc5e900a0b/src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs get only objects from _valueNames that is a list of brackets variables in the formatted string.  For that reason I implemented my own extended versions:

```c#
    public class FormattedLogValuesExposed : FormattedLogValues
    {
        public readonly object[] Parameters;
        public FormattedLogValuesExposed(string format, params object[] values) : base(format, values)
        {
            Parameters = values;
        }
    }
```

Then I only need to use it like: 

```c#        
public static void LogTraceAsync(this ILogger logger, string message, Exception exception, EventId eventId, params object[] args)
        {
            Contract.Requires<ArgumentNullException>(logger == null, "Logger cannot be null.");

            Task.Run(() =>
            {
                logger.Log<FormattedLogValuesExposed>(
                    LogLevel.Trace,
                    eventId,
                    new FormattedLogValuesExposed(message, args),
                    exception,
                    (state,ex) => state.ToString() );

            }).ConfigureAwait(false);
        }
```
and on my Ilogger implementation I just do:

```c#

        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
        {
            if (IsEnabled(logLevel))
            {
                Contract.Requires<ArgumentNullException>(formatter == null, $"Formatter: {nameof(formatter)} cannot be null");

                var message = formatter(state, exception);

                var stateExposed = state as FormattedLogValuesExposed;

                if (stateExposed != null )
                {
                    var json = JsonConvert.SerializeObject(stateExposed.Parameters);

                    // log parameters into db
                }
            }
        }

```

Hope they implement something like this.

**Comment by roji** [comment]
@Ralstlin yeah, this is pretty much what @herecydev proposed [here](https://github.com/aspnet/Logging/issues/533#issuecomment-268217469). The problem, as I wrote further down, is that you simply have an unnamed list of object parameters, whereas in structured logging you usually want to attach names to them - and that's not possible. The real solution here would be to simply pass a `Dictionary<string,object>`, but that means dumping the standard extension methods (e.g. LogDebug).

**Comment by herecydev** [comment]
> but that means dumping the standard extension methods (e.g. LogDebug).

I think you hit the nail on the head. It's not that you couldn't write extensions to fulfill what you need. I think the biggest problem, is that you basically have to tell people not to use those convenient extensions; it's unintuitive. Following the principle of least astonishment, if the out-of-the-box extensions don't work it makes it really hard to get on board.

**Comment by roji** [comment]
Yeah. To be fair, having structured data that isn't referenced in the string message may not be something everyone needs - it's OK to have other Log* extension methods with dictionary overloads.

What I'd like to know is whether a pull request adding these dictionary-using extensions would be accepted...

**Comment by herecydev** [comment]
I don't think they would detract from what's there. I would imagine the official standpoint would have to consider maintenance overhead. I would like to see this included though.

**Comment by Ralstlin** [comment]
I see what you mean, because I am using a Fire and Forget approach for my Async logging, I really could use reflection to get a Name, but it have too many complex issues (same type logged twice, Anonymous types, etc) I modified the previous code to use KeyValuePair, is not as clean as before but is not as bad to discard the option. I prefer to work directly with KeyValuePair because I dont see the benefit of having uniques identifiers for logging.

**Comment by Ralstlin** [comment]
Another option could be to request _values to become a public field. https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/Internal/FormattedLogValues.cs that way these interested in ALL params will have a way to access all them via TState

**Comment by brettsam** [comment]
I wanted to chime in here -- I was looking for this same functionality and came across this issue. I created this class and thought it may be useful: https://gist.github.com/brettsam/baf21619b280912159b4178650294fcd.

My `ILogger`s don't care whether they're getting a `FormattedLogValues` or my custom class -- in both cases they're able to:
- call `ToString()` to get a formatted message
- cast to `IEnumerable<KeyValuePair<string, object>>` to get all of the key-value pairs.

I could write some `LogInfo`/`LogDebug`/etc extensions to make it even easier, but here's how I use it:

    LogValueCollection payload = new LogValueCollection("logging {key}", new[] { "value" }, someDictionaryOfAdditionalProperties);
    logger.Log(LogLevel.Information, 0, payload, null, (s, e) => s.ToString());

**Comment by AlanBarber** [comment]
+1 to running into this issue while working on a custom logger.

As a workaround, I found that I could cast the state as a FormatttedLogValues object and then run a foreach over it to get a the properties attached.

```
public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
{
	var formattedLogValues = state as FormattedLogValues;
	if (formattedLogValues != null)
	{
		foreach (var item in formattedLogValues)
		{
			// Do stuff with item.Key & item.Value
		}
	}
...
}
```

**Comment by jaredcnance** [comment]
what's the status of this request? i'd also be happy to work on a PR, based on @roji's suggestions, if the proposal would be accepted. Although, I'm not sure it needs to be a `Dictionary`, why not just pass `params object[] args` down to the logger? This would allow things like:
```csharp
_log.LogInformation("stuff happened", new { clientId } );
```
and an `ILogger` may handle that however it chooses

**Comment by davidfowl** [comment]
Seems like a reasonable request. I like https://github.com/aspnet/Logging/issues/533#issuecomment-285690699 but I think adding extension methods might actually confuse things unless you had to pass this type explicitly. Would something like this be supported?:

```C#
_log.LogInformation("Something {property}", p1, new { property2 = p2 });
```

Would the extension method combine the properties in the template with the extra properties passed in?

Just FYI, when you add scopes with properties, loggers are supposed to do the same thing, that is, combine properties from the ambient scope into all existing log messages. Good logger implementations should do this (some of ours don't 😞).

Do you have any thoughts on this @nblumhardt?

/cc @glennc @pakrym

**Comment by pakrym** [comment]
Having feature is nice but we need a way to do it without allocations.

**Comment by davidfowl** [comment]
@pakrym suggested we can invent a new syntax for naming the positional arguments without printing them in the log message. This saves the dictionary allocation and allows you to use the precompiled templates.

```C#
_log.LogInformation("Foo {foo}[bar,baz]", foo, bar, baz);
```

**Comment by jaredcnance** [comment]
@davidfowl how does the proposed format differ from:

```csharp
_log.LogInformation("Foo {foo}{bar}{baz}", foo, bar, baz);
```
since you're still required to include the parameters in the message. i still think that the decision about what should be done with `params object[] args` should be delegated to the `ILogger` implementation. but maybe i'm alone on that.

**Comment by nblumhardt** [comment]
I think extending the syntax is probably a fresh can of worms :-) .. it would achieve the no-alloc goal, but there's a lot of design space, docs, and integration work lurking there.

(We've pushed back against adding much more to templates than just "named placeholders" because additional syntax will work against discoverability and reliability in the long run. It's tough to remember special-case syntax for just one or two log messages in an app, and likely that syntax mistakes in message templates won't be caught at development time.)

Something like `BeginScope()` seems like a simpler way to go. I'll propose `WithProperty()` as a similar API without the ambient state/scope stack involved.

First, to _make additional properties available_ without requiring allocations in the calling code, we could use:

```csharp
var scope = log.WithProperty("SomeProperty", someValue);
scope.LogInformation("This has the property attached");
```

Where the value argument is generic `T`.

Now, if the logger is enabled, this still necessitates an allocation for the return value, but levelling `WithProperty()` would fix that:

```csharp
var scope = log.WithProperty(LogLevel.Debug, "SomeProperty", someValue);
// If debug level events aren't enabled, scope == log and no property is attached
```

**Comment by pakrym** [comment]
@nblumhardt it's allocation per property value for loggers that are enabled, a bit too much for a common scenario,

**Comment by nblumhardt** [comment]
@pakrym fair point 👍 

Just so I can understand the cost properly... any formatted event that goes through the pipeline today requires allocations for the state object and the arguments - is the cost here much different?

**Comment by akiander** [comment]
@roji - a sincere thank you for identifying this issue. I am trying to implement an ILogger and I have this exact problem. I opened up #655 for this issue, which can be closed since this issue covers it. 

Because the field is private, I had to use reflection to get to the raw values passed to the logger: 

    var field = typeof(FormattedLogValues).GetRuntimeFields().Where(f => f.IsPrivate && !f.IsStatic && f.Name == "_values").FirstOrDefault();

But that is not an appropriate solution, just a temporary workaround until this issue is solved correctly. 

In my opinion, I don't think any syntax change is required here. I think this issue is solved by making the `_values` field public (readonly): 

[Microsoft.Extensions.Logging.Internal.FormattedLogValues._values](https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Abstractions/Internal/FormattedLogValues.cs#L23)

That way no matter what gets passed, the ILogger can access it.

**Comment by muratg** [comment]
cc @glennc for his thoughts on this.

**Comment by StephenCleary** [comment]
@davidfowl 
> when you add scopes with properties, loggers are supposed to do the same thing, that is, combine properties from the ambient scope into all existing log messages.

Do you mean by this that a log message template can reference parameters defined in its enclosing scopes, or just that the scope can/should be output by the logger? I don't see any loggers doing the former.

**Comment by AddictivKaos** [comment]
I like the idea of leaving methods and syntax as is. If the trailing parameter(s) don't have placeholders in the message template, then I'd like them to be logged automatically as structured data.

One downside to this approach is that you don't get names. And someone mentioned you don't want to use reflection to get the name, but isn't reflection used by the loggers like NLog to serialise the properties of the objects passed as parameters (noting caching is used for Types, but mightn't be advantageous for variable name lookup)?  
If reflection is completely out as an option, what about just auto-assigning a unique name? Similar to anonymous type names or something similar? Could you take advantage of the order of parameters and use the index or offset? Like "__ExtraData0__" : "value", "__ExtraData1__" : 22, etc?

**Comment by andycmaj** [comment]
I wanna jump in here for a sec because i think i might be able to clarify the actual use case the OP was thinking of.

templated message logs are great, but when you're piping your log events to an index like splunk, stackdriver, etc. you really don't need any message at all. When you're trying to consume application logs, i find the context/DATA is the most important thing. I don't care about an actual human-readable message, eg `"A thing happened during {step}"`. In real-world scenarios where developers are looking at tons of log events, they want to be able to data-mine, and thousands of events, all with the same english sentence as a message, are mostly noise.

I think of my log events as events. `EventName`, `Data` (and other metadata like `Source`, `Level`, etc.).

So the logging API i've wrapped around serilog, which has the same formatted-message API, looks like this:

```csharp
eventLogger.InfoEvent(string eventName, object eventData);

eventLogger.InfoEvent("Foo", new { PropA = "bar", PropB = 42 });
```

and the events just look like

```json
{
  "EventName": "Foo",
  "Data": {
    "PropA": "bar",
    "PropB": 42
  }
}
```

So yeah, in production, nobody is looking at the console output of my processes, i'm streaming logs into an index. Once they're there it's kinda pointless to have a "message". 

That said, i can certainly see where framework-level logging might want longer messaging that helps the user take actions and diagnose problems. But for application-level logging, i just need to know what happened, what was the context.

So yeah, unique event names and data bags. no human-readable sentences.

**Comment by andycmaj** [comment]
If anyone is still following this issue, i have a structuredeventlogger library here: https://github.com/andycmaj/SerilogEventLogger. ([nuget](https://www.nuget.org/packages/SerilogEventLogger/))

it uses Serilog as a backend and can integrate with native AspNetCore logging via [Serilog.AspNetCore](https://github.com/serilog/serilog-aspnetcore)

```
// Redirect framework logging to Serilog logger
services.AddSingleton<ILoggerFactory>(_ => new SerilogLoggerFactory(null, true));
```

this implements the API i mention in the previous comment.
